Time-Card/TOOL/main.go

```diff
@@ -0,0 +1,146 @@
+package main
+
+import (
+	"bytes"
+	"encoding/binary"
+	"flag"
+	"fmt"
+	"os"
+
+	crc "github.com/sigurn/crc16"
+	log "github.com/sirupsen/logrus"
+)
+
+type Config struct {
+	Apply            bool
+	InputPath        string
+	OutputPath       string
+	VendorId         int
+	DeviceId         int
+	HardwareRevision int
+}
+
+type Header struct {
+	Magic            [4]byte
+	VendorId         uint16
+	DeviceId         uint16
+	ImageSize        uint32
+	HardwareRevision uint16
+	CRC              uint16
+}
+
+func main() {
+	c := &Config{}
+
+	var logLevel string
+	flag.BoolVar(&c.Apply, "apply", false, "Create new firmware file with the header in the beginning")
+	flag.StringVar(&c.InputPath, "input", "", "Path to raw firmware file")
+	flag.StringVar(&c.OutputPath, "output", "", "Path to firmware file with header (will be overwritten")
+	flag.IntVar(&c.VendorId, "vendor", 0, "PCI VEN_ID to add to header")
+	flag.IntVar(&c.DeviceId, "device", 0, "PCI DEV_ID to add to header")
+	flag.IntVar(&c.HardwareRevision, "hw", 0, "PCI REV_ID (Hardware revision) to add to header")
+	flag.StringVar(&logLevel, "loglevel", "info", "Set a log level. Can be: debug, info, warning, error")
+	flag.Parse()
+
+	switch logLevel {
+	case "debug":
+		log.SetLevel(log.DebugLevel)
+	case "info":
+		log.SetLevel(log.InfoLevel)
+	case "warning":
+		log.SetLevel(log.WarnLevel)
+	case "error":
+		log.SetLevel(log.ErrorLevel)
+	default:
+		log.Fatalf("Unrecognized log level: %v", logLevel)
+	}
+
+	if c.InputPath == "" {
+		log.Fatalf("No input file provided")
+	}
+
+	if c.Apply && c.OutputPath == "" {
+		log.Fatalf("No output file provided")
+	}
+
+	if c.VendorId <= 0 || c.VendorId > 65535 {
+		log.Fatalf("Empty or incorrect PCI Vendor ID specified")
+	}
+
+	if c.DeviceId <= 0 || c.DeviceId > 65535 {
+		log.Fatalf("Empty or incorrect PCI Device ID specified")
+	}
+
+	if c.HardwareRevision < 0 || c.HardwareRevision > 65535 {
+		log.Fatalf("Incorrect PCI Device Revision ID specified")
+	}
+
+	hdr := &Header{
+		Magic:            [4]byte{'O', 'C', 'T', 'C'},
+		VendorId:         uint16(c.VendorId),
+		DeviceId:         uint16(c.DeviceId),
+		HardwareRevision: uint16(c.HardwareRevision),
+	}
+
+	inputFile, err := os.Open(c.InputPath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer inputFile.Close()
+	stat, err := inputFile.Stat()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	hdr.ImageSize = uint32(stat.Size())
+
+	var outputFile *os.File
+	if c.Apply {
+		outputFile, err = os.Create(c.OutputPath)
+		if err != nil {
+			log.Fatal(err)
+		}
+		h := new(bytes.Buffer)
+		binary.Write(h, binary.BigEndian, &hdr)
+		_, err = outputFile.Write(h.Bytes())
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
```

Comment:
This should all go to a separate function and covered with some very basic tests.